### PR TITLE
Add Setoid typeclass

### PR DESCRIPTION
**Pull request by harmenjanssen**
Enables objects to be used with the equals() function.

This also adds a trait for asserting your own implementations obey the Setoid laws.
Reference: https://github.com/fantasyland/fantasy-land#setoid

Note that I need to add some documentation for this, still.

### CODE REVIEW

**Comment by harmenjanssen** [comment]
Note to self: look into https://github.com/steos/php-quickcheck for the TestsSetoidLaws trait.

**Review by martijngastkemper — APPROVED** [review]
LGTM!

File: docs/index.md [hunk 1]
```diff
@@ -1478,3 +1499,47 @@ f\zip($miles, $john); // [
...
 
+## Typeclasses
+
+`Garp\Functional` implements various typeclasses, following [The Fantasyland Specification](https://github.com/fantasyland/fantasy-land).  
+Implementing these takes time, which means this list won't be complete for a while. Pull Requests are very welcome.
+
+A typeclass is an algebraic datatype that adheres to some laws.  
+The specific implementation is up to you, as long as the laws are obeyed, the object can be used with functions in this library, and objects can be composed to form new functionality.
+
+Read the [Fantasyland Specification](https://github.com/fantasyland/fantasy-land) for in-depth details, and [Tom Harding's Fantas, Eel, And Specification](http://www.tomharding.me/fantasy-land/) for concrete implementations (in Javascript) of these types and why they're so incredibly useful.
```
Comment (martijngastkemper) [inline docs/index.md]: This link is already added in the first paragraph. The link to Tom Harding is enough in this paragraph in my opinion. You say "Read the..." so I would expect a peach of text not a Git repo.

Comment (harmenjanssen) [reply docs/index.md]: Fair enough. 🙂 
I will tweak the text!

**Pull request merged**